src/Workspaces/Core/Portable/Remote/RemoteHostSessions.cs

```diff
@@ -0,0 +1,280 @@
+                await session.InvokeAsync(targetName, arguments).ConfigureAwait(false);
+                return true;
+            }
+        }
+
+        public async Task<(bool success, T result)> TryInvokeAsync<T>(string targetName, Solution solution, params object[] arguments)
+        {
+            using (await _gate.DisposableWaitAsync(_cancellationToken).ConfigureAwait(false))
+            using (var scope = await solution.GetPinnedScopeAsync(_cancellationToken).ConfigureAwait(false))
+            {
+                var session = await TryGetSession_NoLockAsync().ConfigureAwait(false);
+                if (session == null)
+                {
+                    return (false, default(T));
+                }
+
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return (true, await session.InvokeAsync<T>(targetName, arguments).ConfigureAwait(false));
+            }
+        }
+
+        public async Task<bool> TryInvokeAsync(
+            string targetName, Solution solution, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync)
+        {
+            using (await _gate.DisposableWaitAsync(_cancellationToken).ConfigureAwait(false))
+            using (var scope = await solution.GetPinnedScopeAsync(_cancellationToken).ConfigureAwait(false))
+            {
+                var session = await TryGetSession_NoLockAsync().ConfigureAwait(false);
+                if (session == null)
+                {
+                    return false;
+                }
+
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                await session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync).ConfigureAwait(false);
+                return true;
+            }
+        }
+
+        public async Task<(bool success, T result)> TryInvokeAsync<T>(string targetName, Solution solution, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync)
+        {
+            using (await _gate.DisposableWaitAsync(_cancellationToken).ConfigureAwait(false))
+            using (var scope = await solution.GetPinnedScopeAsync(_cancellationToken).ConfigureAwait(false))
+            {
+                var session = await TryGetSession_NoLockAsync().ConfigureAwait(false);
+                if (session == null)
+                {
+                    return (false, default(T));
+                }
+
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return (true, await session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync).ConfigureAwait(false));
+            }
+        }
+
+        private async Task<RemoteHostClient.Session> TryGetSession_NoLockAsync()
+        {
+            if (_session != null)
```

Comment:
This code seems to indicate that client/session can be null in the constructor.  if so, can you mark them as 'opt'.  if not, i can't understand how this code woudl ever fire.